### ISSUE TITLE
Task containers dependencies resolution stuck forever

### ISSUE THREAD

**Issue by Trane9991** [aws/amazon-ecs-agent#2579]
### Summary
Task stuck in **Pending** state because of some problems during containers dependencies resolution.
Setting `Start timeout` doesn't have any effect as well.
This is reopening of my previous issue #2350, with the additional details and a Task Definition to reproduce the error.

### Description
I have 3 containers:

| Container name | Essential | Depends On Container | Depends On State | Start Timeout |
|----------------|-----------|----------------------|------------------|---------------|
| exit1          | false     |                      |                  | 10            |
| mysql          | true      | exit1                | SUCCESS          | 30            |
| nginx          | true      | mysql                | START            | 60            |

(`nginx` depends on `mysql` which depends on `exit1`)

When container `exit1` fails with the exit code `1`, containers `mysql` and `nginx` remains in the **PENDING** state forever.
When container `exit1` succeeds with exit code `0` everything works fine.

I also tried to set container `nginx` dependencies like:
```json 
      [
        {
          "containerName": "mysql",
          "condition": "START"
        },
        {
          "containerName": "exit1",
          "condition": "SUCCESS"
        }
      ]
```
the result is the same.

However, behavior seems to be correct if the order of dependencies for `nginx` is changed:
```json 
      [
        {
          "containerName": "exit1",
          "condition": "SUCCESS"
        },
        {
          "containerName": "mysql",
          "condition": "START"
        }
      ]
```

### Expected Behavior
The task is failed to start.

### Observed Behavior
Task stuck in the **PENDING** state forever

<img width="1106" alt="Screen Shot 2020-08-18 at 5 02 17 PM" src="https://user-images.githubusercontent.com/9948629/90522648-a346bf00-e174-11ea-98a8-b07534ae04c6.png">

### Environment Details
- ECS Agent version: 1.43.0

### Task Definitions to reproduce:
<details>
  <summary> With transitive dependencies - stuck as <b>PENDING</b> ❌ </summary>

```json
{
    "containerDefinitions": [
        {
            "command": [
                "cat",
                "123"
            ],
            "image": "alpine",
            "startTimeout": 10,
            "name": "exit1",
            "essential": false
        },
        {
            "image": "nginx",
            "startTimeout": 60,
            "dependsOn": [
                {
                    "containerName": "mysql",
                    "condition": "START"
                }
            ],
            "name": "nginx"
        },
        {
            "image": "mysql:5.7",
            "startTimeout": 30,
            "dependsOn": [
                {
                    "containerName": "exit1",
                    "condition": "SUCCESS"
                }
            ],
            "name": "mysql"
        }
    ],
    "memory": "100",
    "family": "reproduce-dependency-problem",
    "requiresCompatibilities": [
        "EC2"
    ],
    "cpu": "128"
}
```
</details>

<details>
  <summary> With unordered multiple dependencies - stuck as <b>PENDING</b> ❌  </summary>

```json
{
    "containerDefinitions": [
        {
            "command": [
                "cat",
                "123"
            ],
            "image": "alpine",
            "startTimeout": 10,
            "name": "exit1",
            "essential": false
        },
        {
            "image": "nginx",
            "startTimeout": 60,
            "dependsOn": [
                {
                    "containerName": "mysql",
                    "condition": "START"
                },
                {
                    "containerName": "exit1",
                    "condition": "SUCCESS"
                }
            ],
            "name": "nginx"
        },
        {
            "image": "mysql:5.7",
            "startTimeout": 30,
            "dependsOn": [
                {
                    "containerName": "exit1",
                    "condition": "SUCCESS"
                }
            ],
            "name": "mysql"
        }
    ],
    "memory": "100",
    "family": "reproduce-dependency-problem",
    "requiresCompatibilities": [
        "EC2"
    ],
    "cpu": "128"
}
```
</details>

<details>
  <summary> With logically ordered multiple dependencies - task <b>STOPPED</b> as expected ✅ </summary>

```json
{
    "containerDefinitions": [
        {
            "command": [
                "cat",
                "123"
            ],
            "image": "alpine",
            "startTimeout": 10,
            "name": "exit1",
            "essential": false
        },
        {
            "image": "nginx",
            "startTimeout": 60,
            "dependsOn": [
                {
                    "containerName": "exit1",
                    "condition": "SUCCESS"
                },
                {
                    "containerName": "mysql",
                    "condition": "START"
                }
            ],
            "name": "nginx"
        },
        {
            "image": "mysql:5.7",
            "startTimeout": 30,
            "dependsOn": [
                {
                    "containerName": "exit1",
                    "condition": "SUCCESS"
                }
            ],
            "name": "mysql"
        }
    ],
    "memory": "100",
    "family": "reproduce-dependency-problem",
    "requiresCompatibilities": [
        "EC2"
    ],
    "cpu": "128"
}
```
</details>

**Comment by shubham2892** [comment]
@Trane9991 Sorry that you are facing this issue, I will try to reproduce this on my end. 

Meanwhile, Is it possible for you to send the task definition, with which you are seeing the issue, to ecs-agent-external@amazon.com.

**Comment by Trane9991** [comment]
Hey @shubham2892 
Thanks for the quick reply, I published the Task Definitions to reproduce the issues under the collapsable spoilers in the **Task Definitions to reproduce** section of the issue description :)

**Comment by Trane9991** [comment]
Let me know if you are able to reproduce that because it reproduces in 100% cases for me with Task Definitions shared in the issue description.

**Comment by shubham2892** [comment]
I was able to reproduce the pending state behavior with `With transitive dependencies` task def and `With unordered multiple dependencies`, will mark this as a bug and work on getting this fixed.

**Comment by ellenthsu** [comment]
This fix has been released as part of ECS Agent 1.44.4: https://github.com/aws/amazon-ecs-agent/releases/tag/1.44.4

Please perform an update of the Agent: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-update.html or you can find the latest ECS Optimized AMIs containing ECS Agent 1.44.4 here: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-optimized_AMI.html

**Comment by lucabelmonte** [comment]
Hi,

![image](https://user-images.githubusercontent.com/12795970/168309030-7a525bf0-c164-4d66-989f-0c03b0783a3b.png)

As you can see in the attached screenshot, i have the same error as mentioned above. 
I'm currently using the agent version 1.60.1.

![image](https://user-images.githubusercontent.com/12795970/168309276-73d82eb2-6da4-4876-8f0c-51425b4b52af.png)

Here my task definition:
```
{
  "ipcMode": null,
  "executionRoleArn": null,
  "containerDefinitions": [
    {
      "dnsSearchDomains": null,
      "environmentFiles": null,
      "logConfiguration": {
        "logDriver": "awslogs",
        "secretOptions": null,
        "options": {
          "awslogs-group": "/ecs/test-sequential-container",
          "awslogs-region": "eu-west-1",
          "awslogs-stream-prefix": "ecs"
        }
      },
      "entryPoint": null,
      "portMappings": [],
      "command": [
        "/bin/bash",
        "-c",
        "echo ciao && sleep 10 && exit 1"
      ],
      "linuxParameters": null,
      "cpu": 0,
      "environment": [],
      "resourceRequirements": null,
      "ulimits": null,
      "dnsServers": null,
      "mountPoints": [],
      "workingDirectory": null,
      "secrets": null,
      "dockerSecurityOptions": null,
      "memory": 128,
      "memoryReservation": null,
      "volumesFrom": [],
      "stopTimeout": null,
      "image": "ubuntu",
      "startTimeout": null,
      "firelensConfiguration": null,
      "dependsOn": null,
      "disableNetworking": null,
      "interactive": null,
      "healthCheck": null,
      "essential": false,
      "links": null,
      "hostname": null,
      "extraHosts": null,
      "pseudoTerminal": null,
      "user": null,
      "readonlyRootFilesystem": null,
      "dockerLabels": null,
      "systemControls": null,
      "privileged": null,
      "name": "container-1"
    },
    {
      "dnsSearchDomains": null,
      "environmentFiles": null,
      "logConfiguration": {
        "logDriver": "awslogs",
        "secretOptions": null,
        "options": {
          "awslogs-group": "/ecs/test-sequential-container",
          "awslogs-region": "eu-west-1",
          "awslogs-stream-prefix": "ecs"
        }
      },
      "entryPoint": null,
      "portMappings": [],
      "command": [
        "/bin/bash",
        "-c",
        "echo container2 && sleep 10"
      ],
      "linuxParameters": null,
      "cpu": 0,
      "environment": [],
      "resourceRequirements": null,
      "ulimits": null,
      "dnsServers": null,
      "mountPoints": [],
      "workingDirectory": null,
      "secrets": null,
      "dockerSecurityOptions": null,
      "memory": 128,
      "memoryReservation": null,
      "volumesFrom": [],
      "stopTimeout": null,
      "image": "ubuntu",
      "startTimeout": 10,
      "firelensConfiguration": null,
      "dependsOn": [
        {
          "containerName": "container-1",
          "condition": "SUCCESS"
        }
      ],
      "disableNetworking": null,
      "interactive": null,
      "healthCheck": null,
      "essential": false,
      "links": null,
      "hostname": null,
      "extraHosts": null,
      "pseudoTerminal": null,
      "user": null,
      "readonlyRootFilesystem": null,
      "dockerLabels": null,
      "systemControls": null,
      "privileged": null,
      "name": "container-2"
    },
    {
      "dnsSearchDomains": null,
      "environmentFiles": null,
      "logConfiguration": {
        "logDriver": "awslogs",
        "secretOptions": null,
        "options": {
          "awslogs-group": "/ecs/test-sequential-container",
          "awslogs-region": "eu-west-1",
          "awslogs-stream-prefix": "ecs"
        }
      },
      "entryPoint": null,
      "portMappings": [],
      "command": [
        "echo",
        "container3"
      ],
      "linuxParameters": null,
      "cpu": 0,
      "environment": [],
      "resourceRequirements": null,
      "ulimits": null,
      "dnsServers": null,
      "mountPoints": [],
      "workingDirectory": null,
      "secrets": null,
      "dockerSecurityOptions": null,
      "memory": 128,
      "memoryReservation": null,
      "volumesFrom": [],
      "stopTimeout": null,
      "image": "ubuntu",
      "startTimeout": null,
      "firelensConfiguration": null,
      "dependsOn": [
        {
          "containerName": "container-2",
          "condition": "SUCCESS"
        }
      ],
      "disableNetworking": null,
      "interactive": null,
      "healthCheck": null,
      "essential": true,
      "links": null,
      "hostname": null,
      "extraHosts": null,
      "pseudoTerminal": null,
      "user": null,
      "readonlyRootFilesystem": null,
      "dockerLabels": null,
      "systemControls": null,
      "privileged": null,
      "name": "container-3"
    }
  ],
  "placementConstraints": [],
  "memory": null,
  "taskRoleArn": null,
  "compatibilities": [
    "EXTERNAL",
    "EC2"
  ],
  "taskDefinitionArn": "arn:aws:ecs:eu-west-1:XXXX:task-definition/test-sequential-container:15",
  "family": "test-sequential-container",
  "requiresAttributes": [
    {
      "targetId": null,
      "targetType": null,
      "value": null,
      "name": "com.amazonaws.ecs.capability.logging-driver.awslogs"
    },
    {
      "targetId": null,
      "targetType": null,
      "value": null,
      "name": "com.amazonaws.ecs.capability.docker-remote-api.1.19"
    },
    {
      "targetId": null,
      "targetType": null,
      "value": null,
      "name": "ecs.capability.container-ordering"
    }
  ],
  "pidMode": null,
  "requiresCompatibilities": [
    "EC2"
  ],
  "networkMode": null,
  "runtimePlatform": null,
  "cpu": null,
  "revision": 15,
  "status": "ACTIVE",
  "inferenceAccelerators": null,
  "proxyConfiguration": null,
  "volumes": []
}
```

What am i missing?

Mentions: @ellenthsu 

Thanks,

**Comment by angelcar** [comment]
Hi! @lucabelmonte
Thanks for reaching out. 

I was able to repro the issue using your task definition. I will re-open this issue and label it as a bug. We'll work to resolve it.

Correct me if I'm wrong, but I think in this particular case the expectation would be that `container-2` would just stop since `SUCCESS` is never going to happen.

**Comment by lucabelmonte** [comment]
Hi @angelocar,
Thank you for your answer.

Yes exactly. I was expecting the second container `container-2` to go into a` STOPPED` state and also the runned task definition (for a broken dependence).
I also noticed that when there are only 2 containers in a chain, it works.

**Comment by angelcar** [comment]
@lucabelmonte, a fix has been merged and will soon be released.

**Comment by gregmoy** [comment]
I wasn't having this problem before but now I am after upgrading from 1.61.1 to 1.61.3...

**Comment by mssrivas** [comment]
The fix was released in 1.61.2

**Comment by jonioni** [comment]
Wondering if this has been applied to Fargate? Encountered similar issue on fargate tasks.
